### ISSUE TITLE
Wrong 'incompatible return type'

### ISSUE THREAD

**Issue by heikoburau** [python/mypy#12989]
**Bug Report**

Mypy thows an error where everything should be alright.

**To Reproduce**

```
from typing import TypeVar

class Hello:
    pass


class World:
    pass

HelloWorld = TypeVar("HelloWorld", Hello, World)

def fn(x: HelloWorld) -> HelloWorld:
    if isinstance(x, Hello):
        return Hello()
    if isinstance(x, World):
        return World()
```

**Expected Behavior**

No issues found.

**Actual Behavior**

Got error in line "return Hello()":

`error: Incompatible return value type (got "Hello", expected "World")`

**Your Environment**

- Mypy version used: 0.961
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
  [mypy]
plugins = numpy.typing.mypy_plugin
show_absolute_path = True

[mypy-numpy]
ignore_errors = True

[mypy-numpy.*]
ignore_errors = True
- Python version used: 3.9.12
- Operating system and version: Ubuntu 22.04

**Comment by thierry-martinez** [comment]
I encountered a very similar error with the following shorter example: I think it is caused by the same problem, but if you prefer, I can fill another issue. I believe that the problem seems to be that even if the test condition gives information about the type of `x`, this information is not propagated to the type variable.
```
import typing

T = typing.TypeVar("T")

def f(x: T) -> T:
    if type(x) is str:
        return ""
    return x
```
In this example, `type(x) is str` implies `T is str`.

**Comment by lannuttia** [comment]
I'm also seeing what I believe to be the same problem. Here's an example:

```python
from types import SimpleNamespace


class A(): ...

class B(): ...

namespace = SimpleNamespace()
namespace.A = A
namespace.B = B

class C():
    def to[T: A | B](self, clazz: type[T]) -> T:
        match clazz:
            case namespace.A:
                return A()
            case namespace.B:
                return B()
        raise NotImplementedError("This method has not been implemented for the argument provided")

for clazz in (A, B):
    print(f"Should be equal: {repr(clazz)} is {type(C().to(clazz))}")

# Prints the following:
#
# Should be equal: <class '__main__.A'> is <class '__main__.A'>
# Should be equal: <class '__main__.B'> is <class '__main__.B'>
```

**Comment by hauntsaninja** [comment]
mypy is erroring correctly; that will not do the right thing if `clazz` is a subclass of A or B
